[Feature] Add Parcel Feature for RoF2 Clients

neckkola opened the pull request:
I believe this is ready for review.

Added the parcel system supported by the RoF2 Client.  
- Select merchants are now parcel enabled across Norrath
- Several rules to control the system.  Enabled by default.
- Default is a 50 parcel limit
- Prune system to auto delete parcels over xxx days.  Disabled by default.  Pruning written to PlayerLogs
- Several # commands, namely #parcels add to allow administrators to send parcels to players.  Allows nodrop/notrade
- PlayerLogs enabled for sending/retrieving parcels, inclusive of #parcels add command.
- Parcel icon status enabled in player window

Utilizes several new opcodes, therefore requires an opcode update for patches_rof2.conf
A few new db_tables.

There is also an optional sql required to enable merchants located in  https://github.com/EQEmu/Server/tree/master/utils/sql/git/optional/2024_04_20_Parcel_Merchants_Updates.sql
The ids within are from a standard peq install.  Please validate against your install.  Also, any regular merchant, with the is_parcel_merchant flag set to 1 should work as a parcel merchant.

Any thoughts/suggestions/concerns, please let me know.

Thanks and enjoy.

fryguy503 commented:
I have been running this for a week and it has been heavily used by our players. So far looks fantastic!

Valorith commented:
Awesome work!

Akkadius commented on world/world_event_scheduler.cpp:
```diff
@@ -57,12 +59,53 @@ void WorldEventScheduler::Process(ZSList *zs_list)
 					safe_delete(pack);
 				}
+				if (RuleB(Parcel, EnableParcelMerchants) && RuleB(Parcel, EnablePruning) && e.event_type == ServerEvents::EVENT_TYPE_PARCEL_PRUNE) {
+					LogScheduler("Parcel Prune Event Reached [{}]", e.event_data.c_str());
```
No need for `.c_str()` here

neckkola replied:
Resolved

Akkadius commented on world/zoneserver.cpp:
```diff
@@ -1729,6 +1729,18 @@ void ZoneServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p) {
+		{
+            auto in = (Parcel_Struct *)pack->pBuffer;
+            if (strlen(in->send_to) == 0) {
+                LogError("ServerOP_ParcelDelivery pack received with incorrect character_id of {}.", in->send_to);
```
`[{}]`

neckkola replied:
Resolved

Akkadius commented on zone/gm_commands/parcels.cpp:
```diff
@@ -0,0 +1,264 @@
+		auto player      = entity_list.GetClientByName(player_name.c_str());
+
+		if (arguments < 2) {
+			c->Message(Chat::White, "Usage: #parcels listmemory [Character Name]");
```
We could return early in this input validation and remove the else below and naturally fall-through

neckkola replied:
Resolved

Akkadius commented on zone/gm_commands/parcels.cpp:
```diff
@@ -0,0 +1,264 @@
+						.c_str());
+				return;
+			}
+			else {
```
We return already when (!player) 

We can remove the else

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+					inst->SetPrice(p.second.quantity);
+					inst->SetCharges(1);
+				} else {
+					inst->SetCharges(p.second.quantity > 0 ? p.second.quantity : 1);
```
You can set this above `if (item->ID == PARCEL_MONEY_ITEM_ID) {` since this is the default and eliminate the `else`

For example

```cpp
inst->SetCharges(p.second.quantity > 0 ? p.second.quantity : 1);
if (item->ID == PARCEL_MONEY_ITEM_ID) {
	inst->SetPrice(p.second.quantity);
	inst->SetCharges(1);
}
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+
+	}
+	if (parcels.size() >= PARCEL_LIMIT) {
+		LogError("Found {} parcels for Character {}.  List truncated to the PARCEL_MAX_ITEMS [{}] + 5.",
```
```cpp
if (parcels.size() >= PARCEL_LIMIT) {
	LogError(
		"Found {} parcels for Character {}.  List truncated to the PARCEL_MAX_ITEMS [{}] + 5.",
		parcels.size(),
		GetCleanName(),
		PARCEL_MAX_ITEMS
	);
	SendParcelStatus();
	return;
}
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+{
+	auto results = ParcelsRepository::GetWhere(
+		database,
+		fmt::format("`to_name` = '{}' AND `slot_id` = '{}' LIMIT 1", parcel_in.send_to, parcel_in.item_slot));
```
Bring the ending `)` onto the next line.

```cpp
auto results = ParcelsRepository::GetWhere(
	database,
	fmt::format("`to_name` = '{}' AND `slot_id` = '{}' LIMIT 1", parcel_in.send_to, parcel_in.item_slot)
);
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+				inst->SetPrice(parcel.quantity);
+				inst->SetCharges(1);
+			} else {
+				inst->SetCharges(parcel.quantity > 0 ? parcel.quantity : 1);
```
This can be 

```cpp
inst->SetCharges(parcel.quantity > 0 ? parcel.quantity : 1);
if (item->ID == PARCEL_MONEY_ITEM_ID) {
	inst->SetPrice(parcel.quantity);
	inst->SetCharges(1);
}
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+					}
+				}
+				else {
+					MessageString(
```
This is short enough to stay on one line 

```cpp
MessageString(Chat::Yellow, PARCEL_INV_FULL, merchant->GetCleanName());
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+						);
+					}
+					else {
+						MessageString(
```
This is short enough to stay on one line

```cpp
MessageString(Chat::Yellow, PARCEL_INV_FULL, merchant->GetCleanName());
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+	if (results.empty()) {
+		return PARCEL_BEGIN_SLOT;
+	}
+	else {
```
This else isn't necessary, we already return early

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+	else {
+		for (uint32 i = PARCEL_BEGIN_SLOT; i <= PARCEL_END_SLOT; i++) {
+			auto it = std::find_if(results.cbegin(), results.cend(), [&](const auto &x) { return x.slot_id == i; });
+			if (it != results.end()) {
+				continue;
+			}
+			else {
+				return i;
+			}
```
This can be simplified

```cpp
if (it == results.end()) {
	return i;
}
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+
+	auto const num_of_parcels = GetParcelCount();
+
+	if (num_of_parcels == 0) {
+		data->status = IconOff;
+	}
+	else if (num_of_parcels > PARCEL_MAX_ITEMS) {
+		data->status = Overlimit;
+	}
+	else {
+		data->status = IconOn;
+	}
```
The default case can be on top and remove the else

```cpp
data->status = IconOn;
if (num_of_parcels == 0) {
	data->status = IconOff;
}
else if (num_of_parcels > PARCEL_MAX_ITEMS) {
	data->status = Overlimit;
}
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+void Client::AddParcel(ParcelsRepository::Parcels parcel) {
+	auto result = ParcelsRepository::InsertOne(database, parcel);
+	if (!result.id) {
+		LogError("Failed to add parcel to database.  From {} to {} item {} quantity {}", parcel.from_name,
```
```cpp
LogError(
	"Failed to add parcel to database. From [{}] to [{}] item [{}] quantity [{}]", 
	parcel.from_name,
	parcel.to_name, 
	parcel.item_id, 
	parcel.quantity
);
```

neckkola replied:
Resolved

Akkadius commented on zone/worldserver.cpp:
```diff
@@ -3852,6 +3852,45 @@ void WorldServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p)
 	}
+	case ServerOP_ParcelDelivery:
+	{
+		auto in = (Parcel_Struct *)pack->pBuffer;
+
+		if (strlen(in->send_to) == 0) {
+            LogError("ServerOP_ParcelDelivery pack received with incorrect character_name of {}.", in->send_to);
+            return;
+        }
+
+		for (auto const &c : entity_list.GetClientList()) {
+            if (strcasecmp(c.second->GetCleanName(), in->send_to) == 0) {
+                c.second->MessageString(Chat::Yellow, PARCEL_DELIVERY_ARRIVED);
+                c.second->SendParcelStatus();
+                if (c.second->GetEngagedWithParcelMerchant()) {
+                    c.second->SendParcel(*in);
+                }
+                return;
+            }
+        }
+
+		break;
+	}
+	case ServerOP_ParcelPrune:
+	{
+        for (auto const &c : entity_list.GetClientList()) {
+            if (c.second->GetEngagedWithParcelMerchant()) {
+                c.second->Message(Chat::Red, "Parcel data has been updated.  Please re-open the Merchant Window.");
+
+                c.second->SetEngagedWithParcelMerchant(false);
+                c.second->DoParcelCancel();
+
+                auto out = new EQApplicationPacket(OP_ShopEndConfirm);
+                c.second->QueuePacket(out);
+                safe_delete(out);
+                return;
+            }
+        }
+        break;
+	}
```
This whole section has strange formatting, this ended up with whitespaces. 

You can see where the whitespaces are versus tabs

![image](https://github.com/EQEmu/Server/assets/3319450/35fe15c7-dc7d-4ff3-a5ab-f794b89fe1d9)

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+				auto free_id = GetInv().FindFreeSlot(false, false);
+				if (CheckLoreConflict(inst->GetItem())) {
+					if (RuleB(Parcel, DeleteOnDuplicate)) {
+						MessageString(
```
Message can be on one line

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+						);
+					}
+					else {
+						MessageString(
```
Message can be on one line

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+						}
+					}
+					else {
+						MessageString(
```
Message can be on one line

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+		uint32 item_id       = parcel_in.parcel_item_id;
+		uint32 item_quantity = p->second.quantity;
+		if (!item_id || !item_quantity) {
+			LogError("Attempt to retrieve parcel with erroneous item id or quantity for client character id {}.",
```
```cpp
LogError(
	"Attempt to retrieve parcel with erroneous item id or quantity for client character id {}.",
	CharacterID()
);
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+				return;
+			}
+
+			MessageString(
```
```cpp
MessageString(
	Chat::Yellow,
	PARCEL_DELIVERY,
	merchant->GetCleanName(),
	money.c_str(),
	send_to_client.at(0).character_name.c_str()
);
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+
+			auto result = ParcelsRepository::InsertOne(database, parcel_out);
+			if (!result.id) {
+				LogError("Failed to add parcel to database.  From {} to {} item {} quantity {}", parcel_out.from_name,
```
```cpp
LogError(
	"Failed to add parcel to database. From [{}] to [{}] item [{}] quantity [{}]", 
	parcel_out.from_name,
	parcel_out.to_name, 
	parcel_out.item_id, 
	parcel_out.quantity
);
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+			auto money = inst->DetermineMoneyStringForParcels(parcel_in->quantity);
+
+			if (send_to_client.at(0).character_name.empty()) {
+				MessageString(
```
```cpp
MessageString(
	Chat::Yellow,
	PARCEL_UNKNOWN_NAME,
	merchant->GetCleanName(),
	parcel_in->send_to,
	inst->GetItem()->Name
);
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+
+			auto result = ParcelsRepository::InsertOne(database, parcel_out);
+			if (!result.id) {
+				LogError("Failed to add parcel to database.  From {} to {} item {} quantity {}", parcel_out.from_name,
```
```cpp
LogError(
	"Failed to add parcel to database. From [{}] to [{}] item [{}] quantity [{}]", 
	parcel_out.from_name,
	parcel_out.to_name, 
	parcel_out.item_id, 
	parcel_out.quantity
);
```

neckkola replied:
Resolved

Akkadius commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,725 @@
+	}
+}
+
+void Client::DoParcelCancel()
+{
+	if (parcel_platinum || parcel_gold || parcel_silver || parcel_copper) {
+		m_pp.platinum 	+= parcel_platinum;
+		m_pp.gold 		+= parcel_gold;
+		m_pp.silver 	+= parcel_silver;
+		m_pp.copper 	+= parcel_copper;
+		parcel_platinum = 0;
+		parcel_gold     = 0;
+		parcel_silver   = 0;
+		parcel_copper   = 0;
+		SaveCurrency();
+		SendMoneyUpdate();
+	}
+}
```
gold has a weird indentation

neckkola replied:
Resolved

Akkadius commented on zone/gm_commands/parcels.cpp:
```diff
@@ -0,0 +1,264 @@
+				return;
+			}
+
+			if (inst->IsStackable()) {
```
`quantity` is the default here so we can remove the `else`

```cpp
quantity = 1;
if (inst->IsStackable()) {
	quantity = quantity > inst->GetItem()->StackSize ? inst->GetItem()->StackSize : (int16) quantity;
}
else if (inst->GetItem()->MaxCharges > 0) {
	quantity = quantity >= inst->GetItem()->MaxCharges ? inst->GetItem()->MaxCharges : (int16) quantity;
}
```

neckkola replied:
Resolved

Akkadius commented on zone/gm_commands/parcels.cpp:
```diff
@@ -0,0 +1,264 @@
+				if (parcels.empty()) {
+					c->Message(Chat::White, fmt::format("No parcels could be found for {}", player_name).c_str());
+				}
+				c->Message(Chat::Yellow, fmt::format("Found {} parcels for {}.", parcels.size(), player_name).c_str());
+				for (auto const &p: parcels) {
+					c->Message(
+						Chat::Yellow, fmt::format(
+							"Slot [{:02}] has item id [{:10}] with quantity [{}].",
+							p.second.slot_id, p.second.item_id, p.second.quantity
+						)
+							.c_str());
+				}
```
```cpp
c->Message(
	Chat::Yellow,
	fmt::format("Found {} parcels for {}.", results.size(), player_name).c_str()
);
for (auto const &p: results) {
	c->Message(
		Chat::Yellow, fmt::format(
			"Slot [{:02}] has item id [{:10}] with quantity [{}].",
			p.slot_id,
			p.item_id,
			p.quantity
		).c_str()
	);
}
```

neckkola replied:
Resolved

Akkadius commented on zone/gm_commands/parcels.cpp:
```diff
@@ -0,0 +1,264 @@
+		auto player      = entity_list.GetClientByName(player_name.c_str());
+
+		if (arguments < 2) {
+			c->Message(Chat::White, "Usage: #parcels listdb [Character Name]");
```
We can return early here and remove the else below

neckkola replied:
Resolved

Akkadius commented on zone/gm_commands/parcels.cpp:
```diff
@@ -0,0 +1,264 @@
+				auto results = ParcelsRepository::GetWhere(
+					database, fmt::format("to_name = '{}' ORDER BY slot_id ASC", player_name.c_str()));
+				if (results.empty()) {
+					c->Message(Chat::White, fmt::format("No parcels could be found for {}", player_name).c_str());
```
We can return early in here and remove the else below

neckkola replied:
Resolved

Akkadius commented on zone/client.h:
```diff
@@ -1853,6 +1882,14 @@ class Client : public Mob
 	bool Trader;
 	bool Buyer;
 	std::string BuyerWelcomeMessage;
+	int32                                            parcel_platinum;
+	int32                                            parcel_gold;
+	int32                                            parcel_silver;
+	int32                                            parcel_copper;
+	int32                                            parcel_count;
+	bool                                             parcel_enabled;
+	bool                                             parcel_merchant_engaged;
+	std::map<uint32, BaseParcelsRepository::Parcels> parcels{};
```
Any private member variables we've been trying to prefix with `m_` when we can

eg `m_parcel_platinum`

Clion can easily refactor these accurately

neckkola replied:
Resolved.  Mental note.  Clion auto refactor - Deselect within text areas.  :-)

Akkadius commented on common/patches/rof2.cpp:
```diff
@@ -5638,7 +5699,13 @@ namespace RoF2
 		memset(&ibs, 0, sizeof(RoF2::structs::ItemBodyStruct));
 
-		ibs.id = item->ID;
+		if (item->ID == PARCEL_MONEY_ITEM_ID) {
```
Declare the default case above and remove the else

```cpp
ibs.id = item->ID;
if (item->ID == PARCEL_MONEY_ITEM_ID) {
	ibs.id = inst->GetSerialNumber();
}
```

neckkola replied:
Resolved

Akkadius commented on common/patches/rof2.cpp:
```diff
@@ -5533,12 +5580,19 @@ namespace RoF2
 		//sprintf(hdr.unknown000, "06e0002Y1W00");
 
-		snprintf(hdr.unknown000, sizeof(hdr.unknown000), "%016d", item->ID);
+		if (packet_type == ItemPacketParcel) {
+			strn0cpy(
+				hdr.unknown000, fmt::format("{:03}PAR{:010}\0", inst->GetMerchantSlot(), item->ID).c_str(),
+				sizeof(hdr.unknown000));
+		}
+		else {
+			snprintf(hdr.unknown000, sizeof(hdr.unknown000), "%016d", item->ID);
+		}
```
Default case can be declared above and remove the else

```cpp
snprintf(hdr.unknown000, sizeof(hdr.unknown000), "%016d", item->ID);
if (packet_type == ItemPacketParcel) {
	strn0cpy(
		hdr.unknown000, 
		fmt::format("{:03}PAR{:010}\0", inst->GetMerchantSlot(), item->ID).c_str(),
		sizeof(hdr.unknown000)
	);
}
```

neckkola replied:
Resolved

Akkadius commented on common/events/player_events.h:
```diff
@@ -976,6 +982,72 @@ namespace PlayerEvent {
 		}
 	};
+
+	struct ParcelRetrieve
+	{
+        uint32      item_id;
+        uint32      quantity;
+        std::string from_player_name;
+        uint32      sent_date;
+
+		// cereal
+		template<class Archive>
+		void serialize(Archive &ar)
+		{
+			ar(
+				CEREAL_NVP(item_id),
+				CEREAL_NVP(quantity),
+				CEREAL_NVP(from_player_name),
+				CEREAL_NVP(sent_date)
+			);
+		}
+	};
```
We spoke about this, these structs have spaces versus tabs so indentation looks off here versus in editor

neckkola replied:
Resolved

Akkadius commented on common/eq_packet_structs.h:
```diff
@@ -2052,12 +2066,75 @@ struct TimeOfDay_Struct {
+	std::string    note;
+	uint32         slot_id;
+
+	template<class Archive>
+	void serialize(Archive &archive)
+	{
+		archive(CEREAL_NVP(packet_type),
+				CEREAL_NVP(serialized_item),
+				CEREAL_NVP(sent_time),
+				CEREAL_NVP(player_name),
+				CEREAL_NVP(note),
+				CEREAL_NVP(slot_id)
+		);
+	}
```
```cpp
template<class Archive>
void serialize(Archive &archive)
{
	archive(
		CEREAL_NVP(packet_type),
		CEREAL_NVP(serialized_item),
		CEREAL_NVP(sent_time),
		CEREAL_NVP(player_name),
		CEREAL_NVP(note),
		CEREAL_NVP(slot_id)
	);
}
```

neckkola replied:
Resolved

Akkadius commented:
This is looking really good @neckkola - I think we can land this one pretty quick. Really nice work!

Kinglykrab commented on zone/client_packet.cpp:
```diff
@@ -4268,6 +4276,13 @@ void Client::Handle_OP_Bug(const EQApplicationPacket *app)
 
 void Client::Handle_OP_Camp(const EQApplicationPacket *app)
 {
+	if (ClientVersion() == EQ::versions::ClientVersion::RoF2 && RuleB (Parcel, EnableParcelMerchants) &&
```
`RuleB(Parcel, EnableParcelMerchants)`

neckkola replied:
Resolved

Kinglykrab commented on zone/client_packet.cpp:
```diff
@@ -4321,6 +4336,12 @@ void Client::Handle_OP_CancelTrade(const EQApplicationPacket *app)
 		trade->Reset();
 	}
+
+	if (ClientVersion() == EQ::versions::ClientVersion::RoF2 && RuleB (Parcel, EnableParcelMerchants)) {
```
`RuleB(Parcel, EnableParcelMerchants)`

neckkola replied:
@Kinglykrab I left this in as I figure the client check is better than a packet and db call that would occur.  I suppose, if everyone is on RoF2, it would be more expensive, though with Ti/UF player bases, it will be less expensive.  Let me know your thoughts.

neckkola replied:
Misunderstoond.  Updated the space item.

Kinglykrab commented on zone/gm_commands/parcels.cpp:
```diff
@@ -0,0 +1,264 @@
+
+			ParcelsRepository::Parcels parcel_out;
+			parcel_out.from_name = c->GetName();
+			parcel_out.note      = note.empty() ? "" : note;
```
Don't think we need to do this, since we're setting to empty if it's empty it shouldn't matter, right?

neckkola replied:
Resolved

Kinglykrab commented:
`Parcels and General Supples` is what's in your data. Should be `Parcels and General Supplies`.

neckkola commented on world/zoneserver.cpp:
```diff
@@ -1729,6 +1729,18 @@ void ZoneServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p) {
+            zoneserver_list.SendPacketToBootedZones(pack);
+
+            break;
+        }
```
Resolved, hopefully.  :-)

neckkola commented:
> `Parcels and General Supples` is what's in your data. Should be `Parcels and General Supplies`.

Roger.  The typo.  Resolved.

Kinglykrab commented on zone/client_packet.cpp:
```diff
@@ -4276,8 +4276,7 @@ void Client::Handle_OP_Bug(const EQApplicationPacket *app)
 void Client::Handle_OP_Camp(const EQApplicationPacket *app)
 {
-	if (ClientVersion() == EQ::versions::ClientVersion::RoF2 && RuleB (Parcel, EnableParcelMerchants) &&
-		GetEngagedWithParcelMerchant()) {
```
Why was this removed? Only thing I noticed that could be changed was pushing `RuleB` up to the `(` like `RuleB(`.

neckkola replied:
My mistake.  Misunderstood.  Updated.

Kinglykrab commented on zone/gm_commands/parcels.cpp:
```diff
@@ -184,30 +203,34 @@ void command_parcels(Client *c, const Seperator *sep)
 			Parcel_Struct ps{};
 			ps.item_slot = parcel_out.slot_id;
-			strn0cpy(ps.send_to, parcel_out.to_name.c_str(), sizeof(ps.send_to));
+			strn0cpy(ps.send_to,parcel_out.to_name.c_str(),sizeof(ps.send_to));
```
Spaces removed?

neckkola replied:
I have been having regular fights with the auto formatting.  A royal PITA in fact.  :-)  When I autoformat, it puts everything on separate lines.  Is the standard separate lines or a single?  I was working under the impression that if it was 80 chars or less, then the auto formatter should leave it on a single line?  @Kinglykrab

neckkola replied:
Updated

Kinglykrab commented on zone/gm_commands/parcels.cpp:
```diff
@@ -184,30 +203,34 @@ void command_parcels(Client *c, const Seperator *sep)
 			auto item = database.GetItem(item_id);
 			if (!item) {
-				c->Message(Chat::Yellow, "Could not find an item with id {}", item_id);
+				c->Message(Chat::Yellow,"Could not find an item with id {}",item_id);
```
Spaces removed?

neckkola replied:
Updated

Kinglykrab commented on zone/gm_commands/parcels.cpp:
```diff
@@ -184,30 +203,34 @@ void command_parcels(Client *c, const Seperator *sep)
+			);
 			if (!inst) {
-				c->Message(Chat::Yellow, "Could not find an item with id {}", item_id);
+				c->Message(Chat::Yellow,"Could not find an item with id {}",item_id);
```
Spaces removed?

neckkola replied:
Updated

Kinglykrab commented on zone/gm_commands/parcels.cpp:
```diff
@@ -245,7 +280,7 @@ void command_parcels(Client *c, const Seperator *sep)
 			Parcel_Struct ps{};
 			ps.item_slot = parcel_out.slot_id;
-			strn0cpy(ps.send_to, parcel_out.to_name.c_str(), sizeof(ps.send_to));
+			strn0cpy(ps.send_to,parcel_out.to_name.c_str(),sizeof(ps.send_to));
```
Spaces removed?

neckkola replied:
Updated

Kinglykrab commented on zone/parcels.cpp:
```diff
@@ -149,32 +154,37 @@ void Client::SendParcel(Parcel_Struct parcel_in)
-		parcel_gold     = 0;
-		parcel_silver   = 0;
-		parcel_copper   = 0;
+	if (m_parcel_platinum || m_parcel_gold || m_parcel_silver ||
+		m_parcel_copper) {
+		m_pp.platinum += m_parcel_platinum;
+		m_pp.gold += m_parcel_gold;
+		m_pp.silver += m_parcel_silver;
+		m_pp.copper += m_parcel_copper;
+		m_parcel_platinum = 0;
+		m_parcel_gold     = 0;
+		m_parcel_silver   = 0;
+		m_parcel_copper   = 0;
```
Formatting.
```cpp
if (
	m_parcel_platinum ||
	m_parcel_gold ||
	m_parcel_silver ||
	m_parcel_copper
) {
	m_pp.platinum += m_parcel_platinum;
	m_pp.gold     += m_parcel_gold;
	m_pp.silver   += m_parcel_silver;
	m_pp.copper   += m_parcel_copper;

	m_parcel_platinum = 0;
	m_parcel_gold     = 0;
	m_parcel_silver   = 0;
	m_parcel_copper   = 0;
```

neckkola replied:
Made the changes though the formatter does not do this for me.  I must have some setting incorrect.

Kinglykrab commented on zone/parcels.cpp:
```diff
@@ -99,7 +101,12 @@ void Client::SendParcel(Parcel_Struct parcel_in)
-		fmt::format("`to_name` = '{}' AND `slot_id` = '{}' LIMIT 1", parcel_in.send_to, parcel_in.item_slot));
+		fmt::format(
+			"`to_name` = '{}' AND `slot_id` = '{}' LIMIT 1",
+			parcel_in.send_to,
```
`Strings::Escape(parcel_in.send_to)` maybe?

neckkola replied:
Updated

Kinglykrab commented on zone/gm_commands/parcels.cpp:
```diff
@@ -69,36 +68,38 @@ void command_parcels(Client *c, const Seperator *sep)
+
+		auto parcels = player->GetParcels();
+		if (parcels.empty()) {
+			c->Message(Chat::White,fmt::format("No parcels could be found for {}",player_name).c_str());
```
Formatting, I'd put some spaces are the very least.

neckkola replied:
Updated

Kinglykrab commented on zone/gm_commands/parcels.cpp:
```diff
@@ -69,36 +68,38 @@ void command_parcels(Client *c, const Seperator *sep)
+			return;
+		}
+
+		c->Message(Chat::Yellow,fmt::format("Found {} parcels for {}.",	parcels.size(),	player_name).c_str());
```
Formatting, I'd put some spaces are the very least.

Kinglykrab commented on common/database/database_update_manifest.cpp:
```diff
@@ -5468,7 +5468,61 @@ DROP PRIMARY KEY,
+				COLLATE='latin1_swedish_ci'
+				ENGINE=InnoDB
+				AUTO_INCREMENT=1;
+				REPLACE INTO `parcel_merchants` (`merchant_id`, `last_name`) VALUES 
+				(202129, 'Parcels and General Supplies'),
+				(3036, 'Parcels and General Supplies'),
+				(394025, 'Parcels and General Supplies'),
+				(75113, 'Parcels and General Supplies'),
+				(49073, 'Parcels and General Supplies'),
+				(41021, 'Parcels and General Supplies'),
+				(40070, 'Parcels and General Supplies'),
+				(106115, 'Parcels and General Supplies'),
+				(55150, 'Parcels and General Supplies'),
+				(9053, 'Parcels and General Supplies'),
+				(382156, 'Parcels and General Supplies'),
+				(1032, 'Parcels and General Supplies'),
+				(155088, 'Parcels and General Supplies'),
+				(23017, 'Parcels and General Supplies'),
+				(61065, 'Parcels and General Supplies'),
+				(29008, 'Parcels and General Supplies'),
+				(67058, 'Parcels and General Supplies'),
+				(54067, 'Parcels and General Supplies'),
+				(19031, 'Parcels and General Supplies'),
+				(50140, 'Parcels and General Supplies');
```
Question: Do we want a dedicated table for this, or should we just create an `is_parcel_merchant` flag in `npc_types`?

neckkola replied:
It is a good question and I can adjust as required.  I would certainly support the addition of the column to the npc table.

neckkola replied:
How would we deal with the differences for expansions for the last name.  If pre RoF2, use General Supplies, and RoF2 with the addition of Parcels?  Any thoughts?

Kinglykrab replied:
I think the way they would handle this is a different NPC ID for that expansion of the game with a different lastname and the `is_parcel_merchant` flag set to `true`.

neckkola replied:
Refactored to add the is_parcel_merchant column to db.  I added the updates to the db manifest for lastname, however these will also need to be added to source db once this PR is approved. (@joligario )  I did not duplicate the npc entries at this point and can do that as the expansion work progresses.

hgtw commented on common/patches/rof2.cpp:
```diff
@@ -1582,27 +1583,86 @@ namespace RoF2
+
+				strncpy(player_name, pms.player_name.c_str(), player_name_length);
+				strncpy(note, pms.note.c_str(), note_length);
+				strncpy(serial, pms.serialized_item.c_str(), serial_length);
```
Are these char buffers/copies even needed? `serial` looks unused, can `player_name` and `note` just be written directly to the output buffer from the `pms` strings?

neckkola replied:
Rewrote this to be more efficient.

hgtw commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,755 @@
+	strn0cpy(data->send_to, parcel.send_to, sizeof(data->send_to));
+
+	worldserver.SendPacket(out);
+}
```
Memory leak on `out` here

neckkola replied:
As noted above, converted to unique_ptr.

hgtw commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,755 @@
+						out->Size(),
+						packet_size
+					);
+					return;
```
Memory leak on `out` and `inst` if this returns

neckkola replied:
As noted above, converted to unique_ptr.

hgtw commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,755 @@
+					out->Size(),
+					packet_size
+				);
+				return;
```
Memory leak on `out` and `inst` if this returns

neckkola replied:
As noted above, converted to unique_ptr.

hgtw commented on zone/gm_commands/parcels.cpp:
```diff
@@ -0,0 +1,299 @@
+				return;
+			}
+
+			auto inst = database.CreateItem(item, 1);
```
`database.CreateItem` creates a new item instance on the heap and gives the caller ownership so this needs to be deleted to avoid a memory leak when done with it. Suggest just wrapping it in a unique_ptr

neckkola replied:
As noted above, converted to unique_ptr.

hgtw commented on zone/gm_commands/parcels.cpp:
```diff
@@ -0,0 +1,299 @@
+				return;
+			}
+
+			auto inst = database.CreateItem(
```
`inst` needs to be deleted

neckkola replied:
As noted above, converted to unique_ptr.

hgtw commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,755 @@
+	for (auto &p: m_parcels) {
+		auto item = database.GetItem(p.second.item_id);
+		if (item) {
+			auto inst = database.CreateItem(item, p.second.quantity);
```
`inst` needs to be deleted

neckkola replied:
As noted above, converted to unique_ptr.

hgtw commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,755 @@
+
+	auto item = database.GetItem(parcel.item_id);
+	if (item) {
+		auto inst = database.CreateItem(item, parcel.quantity);
```
`inst` needs to be deleted

neckkola replied:
As noted above, converted to unique_ptr.

hgtw commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,755 @@
+				return;
+			}
+
+			auto inst = database.CreateItem(item, 1);
```
`inst` needs to be deleted when finished with it

hgtw commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,755 @@
+			return;
+		}
+
+		auto inst = database.CreateItem(item_id, item_quantity);
```
`inst` needs to be deleted. Even though `TryStacking` and `PutItemInInventory` are called with this instance below, those functions do not take ownership and just clones it, so this still needs to be freed after

neckkola replied:
As noted above, converted to unique_ptr.

neckkola commented on common/patches/rof2.cpp:
```diff
@@ -1582,27 +1583,86 @@ namespace RoF2
+					LogNetcode("RoF2::ENCODE(OP_ItemPacket) Serialization failed on item slot [{}]", pms.slot_id);
+
+					safe_delete(in);
+					return;
```
Thanks for the comments hg.  I have migrated all the new/deletes to unique_ptr/equivalents.  Thanks again for the advice.  Can you please take a look and make sure they look ok?  Anything else re newer techniques to do memory mgmt, please let me know.  I will no longer use the new/delete construct.

neckkola commented on common/database/database_update_manifest.cpp:
```diff
@@ -5468,7 +5468,61 @@ DROP PRIMARY KEY,
+				`item_id` INT(10) UNSIGNED NOT NULL DEFAULT '0',
+				`slot_id` INT(10) UNSIGNED NOT NULL DEFAULT '0',
+				`quantity` INT(10) UNSIGNED NOT NULL DEFAULT '0',
+				`to_name` VARCHAR(64) NULL DEFAULT NULL COLLATE 'latin1_swedish_ci',
```
Refactored the parcel table to character_parcels as descibed.

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -416,6 +430,13 @@ class BaseNpcTypesRepository {
 			"heroic_strikethrough",
 			"faction_amount",
 			"keeps_sold_items",
+			"chesttexture",
+			"combat_hp_regen",
+			"combat_mana_regen",
+			"aggro_pc",
+			"ignore_distance",
+			"disable_instance",
```
Are these custom columns?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -583,6 +604,13 @@ class BaseNpcTypesRepository {
 		e.heroic_strikethrough   = 0;
 		e.faction_amount         = 0;
 		e.keeps_sold_items       = 1;
+		e.chesttexture           = 0;
+		e.combat_hp_regen        = 0;
+		e.combat_mana_regen      = 0;
+		e.aggro_pc               = 0;
+		e.ignore_distance        = 1;
+		e.disable_instance       = 1;
```
Are these custom columns?

fryguy503 replied:
I think this was from a copy of my database when he was testing things at one point.

neckkola replied:
Yes, that is correct @fryguy503 .  I will revert that and make a commit.  This is the cause of all the other comments you just made as well @Kinglykrab  It was due to the recreate of the repo files after I added the is_parcel_merchant with the other db table.

neckkola replied:
Just pushed a commit on this.  Hope I didn't give you a heart attack @Kinglykrab   Should resolve all the other comments as well.

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -665,8 +693,8 @@ class BaseNpcTypesRepository {
 			e.armortint_blue         = row[45] ? static_cast<uint8_t>(strtoul(row[45], nullptr, 10)) : 0;
-			e.d_melee_texture1       = row[46] ? static_cast<uint32_t>(strtoul(row[46], nullptr, 10)) : 0;
-			e.d_melee_texture2       = row[47] ? static_cast<uint32_t>(strtoul(row[47], nullptr, 10)) : 0;
+			e.d_melee_texture1       = row[46] ? static_cast<int32_t>(atoi(row[46])) : 0;
+			e.d_melee_texture2       = row[47] ? static_cast<int32_t>(atoi(row[47])) : 0;
```
Why was this changed?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -746,6 +774,13 @@ class BaseNpcTypesRepository {
 			e.heroic_strikethrough   = row[124] ? static_cast<int32_t>(atoi(row[124])) : 0;
 			e.faction_amount         = row[125] ? static_cast<int32_t>(atoi(row[125])) : 0;
 			e.keeps_sold_items       = row[126] ? static_cast<uint8_t>(strtoul(row[126], nullptr, 10)) : 1;
+			e.chesttexture           = row[127] ? static_cast<int32_t>(atoi(row[127])) : 0;
+			e.combat_hp_regen        = row[128] ? static_cast<int32_t>(atoi(row[128])) : 0;
+			e.combat_mana_regen      = row[129] ? static_cast<int32_t>(atoi(row[129])) : 0;
+			e.aggro_pc               = row[130] ? static_cast<int32_t>(atoi(row[130])) : 0;
+			e.ignore_distance        = row[131] ? static_cast<int32_t>(atoi(row[131])) : 1;
+			e.disable_instance       = row[132] ? static_cast<int32_t>(atoi(row[132])) : 1;
```
Are these custom columns?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -905,6 +940,13 @@ class BaseNpcTypesRepository {
 		v.push_back(columns[124] + " = " + std::to_string(e.heroic_strikethrough));
 		v.push_back(columns[125] + " = " + std::to_string(e.faction_amount));
 		v.push_back(columns[126] + " = " + std::to_string(e.keeps_sold_items));
+		v.push_back(columns[127] + " = " + std::to_string(e.chesttexture));
+		v.push_back(columns[128] + " = " + std::to_string(e.combat_hp_regen));
+		v.push_back(columns[129] + " = " + std::to_string(e.combat_mana_regen));
+		v.push_back(columns[130] + " = " + std::to_string(e.aggro_pc));
+		v.push_back(columns[131] + " = " + std::to_string(e.ignore_distance));
+		v.push_back(columns[132] + " = " + std::to_string(e.disable_instance));
```
Are these custom columns?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -1053,6 +1095,13 @@ class BaseNpcTypesRepository {
 		v.push_back(std::to_string(e.heroic_strikethrough));
 		v.push_back(std::to_string(e.faction_amount));
 		v.push_back(std::to_string(e.keeps_sold_items));
+		v.push_back(std::to_string(e.chesttexture));
+		v.push_back(std::to_string(e.combat_hp_regen));
+		v.push_back(std::to_string(e.combat_mana_regen));
+		v.push_back(std::to_string(e.aggro_pc));
+		v.push_back(std::to_string(e.ignore_distance));
+		v.push_back(std::to_string(e.disable_instance));
```
Are these custom columns?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -1209,6 +1258,13 @@ class BaseNpcTypesRepository {
 			v.push_back(std::to_string(e.heroic_strikethrough));
 			v.push_back(std::to_string(e.faction_amount));
 			v.push_back(std::to_string(e.keeps_sold_items));
+			v.push_back(std::to_string(e.chesttexture));
+			v.push_back(std::to_string(e.combat_hp_regen));
+			v.push_back(std::to_string(e.combat_mana_regen));
+			v.push_back(std::to_string(e.aggro_pc));
+			v.push_back(std::to_string(e.ignore_distance));
+			v.push_back(std::to_string(e.disable_instance));
```
Are these custom columns?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -1288,8 +1344,8 @@ class BaseNpcTypesRepository {
 			e.armortint_blue         = row[45] ? static_cast<uint8_t>(strtoul(row[45], nullptr, 10)) : 0;
-			e.d_melee_texture1       = row[46] ? static_cast<uint32_t>(strtoul(row[46], nullptr, 10)) : 0;
-			e.d_melee_texture2       = row[47] ? static_cast<uint32_t>(strtoul(row[47], nullptr, 10)) : 0;
+			e.d_melee_texture1       = row[46] ? static_cast<int32_t>(atoi(row[46])) : 0;
+			e.d_melee_texture2       = row[47] ? static_cast<int32_t>(atoi(row[47])) : 0;
```
Why were these changed?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -1369,6 +1425,13 @@ class BaseNpcTypesRepository {
 			e.heroic_strikethrough   = row[124] ? static_cast<int32_t>(atoi(row[124])) : 0;
 			e.faction_amount         = row[125] ? static_cast<int32_t>(atoi(row[125])) : 0;
 			e.keeps_sold_items       = row[126] ? static_cast<uint8_t>(strtoul(row[126], nullptr, 10)) : 1;
+			e.chesttexture           = row[127] ? static_cast<int32_t>(atoi(row[127])) : 0;
+			e.combat_hp_regen        = row[128] ? static_cast<int32_t>(atoi(row[128])) : 0;
+			e.combat_mana_regen      = row[129] ? static_cast<int32_t>(atoi(row[129])) : 0;
+			e.aggro_pc               = row[130] ? static_cast<int32_t>(atoi(row[130])) : 0;
+			e.ignore_distance        = row[131] ? static_cast<int32_t>(atoi(row[131])) : 1;
+			e.disable_instance       = row[132] ? static_cast<int32_t>(atoi(row[132])) : 1;
```
Are these custom columns?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -1439,8 +1502,8 @@ class BaseNpcTypesRepository {
 			e.armortint_blue         = row[45] ? static_cast<uint8_t>(strtoul(row[45], nullptr, 10)) : 0;
-			e.d_melee_texture1       = row[46] ? static_cast<uint32_t>(strtoul(row[46], nullptr, 10)) : 0;
-			e.d_melee_texture2       = row[47] ? static_cast<uint32_t>(strtoul(row[47], nullptr, 10)) : 0;
+			e.d_melee_texture1       = row[46] ? static_cast<int32_t>(atoi(row[46])) : 0;
+			e.d_melee_texture2       = row[47] ? static_cast<int32_t>(atoi(row[47])) : 0;
```
Why were these changed?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -1520,6 +1583,13 @@ class BaseNpcTypesRepository {
 			e.heroic_strikethrough   = row[124] ? static_cast<int32_t>(atoi(row[124])) : 0;
 			e.faction_amount         = row[125] ? static_cast<int32_t>(atoi(row[125])) : 0;
 			e.keeps_sold_items       = row[126] ? static_cast<uint8_t>(strtoul(row[126], nullptr, 10)) : 1;
+			e.chesttexture           = row[127] ? static_cast<int32_t>(atoi(row[127])) : 0;
+			e.combat_hp_regen        = row[128] ? static_cast<int32_t>(atoi(row[128])) : 0;
+			e.combat_mana_regen      = row[129] ? static_cast<int32_t>(atoi(row[129])) : 0;
+			e.aggro_pc               = row[130] ? static_cast<int32_t>(atoi(row[130])) : 0;
+			e.ignore_distance        = row[131] ? static_cast<int32_t>(atoi(row[131])) : 1;
+			e.disable_instance       = row[132] ? static_cast<int32_t>(atoi(row[132])) : 1;
```
Are these custom columns?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -1721,6 +1791,13 @@ class BaseNpcTypesRepository {
 		v.push_back(std::to_string(e.heroic_strikethrough));
 		v.push_back(std::to_string(e.faction_amount));
 		v.push_back(std::to_string(e.keeps_sold_items));
+		v.push_back(std::to_string(e.chesttexture));
+		v.push_back(std::to_string(e.combat_hp_regen));
+		v.push_back(std::to_string(e.combat_mana_regen));
+		v.push_back(std::to_string(e.aggro_pc));
+		v.push_back(std::to_string(e.ignore_distance));
+		v.push_back(std::to_string(e.disable_instance));
```
Are these custom columns?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -1870,6 +1947,13 @@ class BaseNpcTypesRepository {
 			v.push_back(std::to_string(e.heroic_strikethrough));
 			v.push_back(std::to_string(e.faction_amount));
 			v.push_back(std::to_string(e.keeps_sold_items));
+			v.push_back(std::to_string(e.chesttexture));
+			v.push_back(std::to_string(e.combat_hp_regen));
+			v.push_back(std::to_string(e.combat_mana_regen));
+			v.push_back(std::to_string(e.aggro_pc));
+			v.push_back(std::to_string(e.ignore_distance));
+			v.push_back(std::to_string(e.disable_instance));
```
Are these custom columns?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -146,6 +146,13 @@ class BaseNpcTypesRepository {
 		int32_t     heroic_strikethrough;
 		int32_t     faction_amount;
 		uint8_t     keeps_sold_items;
+		int32_t     chesttexture;
+		int32_t     combat_hp_regen;
+		int32_t     combat_mana_regen;
+		int32_t     aggro_pc;
+		int32_t     ignore_distance;
+		int32_t     disable_instance;
```
Are these custom columns?

Kinglykrab commented on common/repositories/base/base_npc_types_repository.h:
```diff
@@ -65,8 +65,8 @@ class BaseNpcTypesRepository {
-		uint32_t    d_melee_texture1;
-		uint32_t    d_melee_texture2;
+		int32_t     d_melee_texture1;
+		int32_t     d_melee_texture2;
```
Why were these changed?

Kinglykrab commented:
I believe some custom changes have found their way into this.

neckkola commented:
> I believe some custom changes have found their way into this.

You believe correctly!  :-)  Should be all fixed up now.

hgtw commented on common/patches/rof2.cpp:
```diff
@@ -36,6 +36,7 @@
 #include "../classes.h"
 #include "../races.h"
 #include "../raid.h"
+#include "../../zone/parcels.h"
```
Probably want to avoid including a zone header in common, does it need a simple struct that can be moved to a common header?

neckkola replied:
Removed parcels.h and place constants within eq_constants.

hgtw commented on zone/parcels.cpp:
```diff
@@ -0,0 +1,755 @@
+	}
+}
+
+void Client::SendParcel(Parcel_Struct parcel_in)
```
Change this to `const Parcel_Struct&` to avoid the copy

neckkola replied:
Update and also checked others and updated as well.

hgtw reviewed (commented):
Leaks look resolved, only real issue is some sql stuff. I didn't deep dive into the implementation but there there might be the possibility of removing the server side string money formatting function in favor of letting the client handle it with the right money item id.

neckkola commented on common/item_instance.h:
```diff
@@ -301,6 +301,7 @@ namespace EQ
 		int GetItemHaste(bool augments = false) const;
 		uint32 GetItemGuildFavor() const;
 		std::vector<uint32> GetAugmentIDs() const;
+		std::string DetermineMoneyStringForParcels(uint32 cp) const;
```
Done

hgtw replied:
Ah, I guess the server side string function would still be needed for any messages that are supposed to show the amount. It was a bad suggestion to remove it completely if it's still needed for those. Otherwise looks fine even if that function needs to be re-added.

neckkola commented on zone/parcels.h:
```diff
@@ -0,0 +1,32 @@
+
+#define PARCEL_SEND_ITEMS    0
+#define PARCEL_SEND_MONEY    1
+#define PARCEL_MONEY_ITEM_ID 22292 // item id of copper coin
```
Did some investigating and got this to work.  Had to make a few other adjustments, though in the end this is working.  Thanks for the tip.

neckkola commented on common/database_schema.h:
```diff
@@ -59,6 +59,7 @@ namespace DatabaseSchema {
 			{"character_leadership_abilities", "id"},
 			{"character_material",             "id"},
 			{"character_memmed_spells",        "id"},
+			{"character_parcels",              "id"},
```
Updated.  Thanks for the point to the other two functions for a better understanding.

neckkola commented on common/database/database_update_manifest.cpp:
```diff
@@ -5468,7 +5468,37 @@ DROP PRIMARY KEY,
+
+				UPDATE `npc_types` SET `is_parcel_merchant` = 1, `lastname` = 'Parcels and General Supplies'
+				WHERE id IN (202129, 3036, 394025, 75113, 49073, 41021, 40070, 106115, 55150, 9053, 382156, 1032,
+				155088, 23017, 61065, 29008, 67058, 54067, 19031, 50140);
```
Agree.  I pinged JJ as he was very helpful with the guild tribute changes to content.  I can adjust as required.

joligario replied:
Sorry, I must have missed or forgot! What do you need from me?

neckkola commented:
Good evening all.  Anything else required for this one?  I will require some assistance with the content items and happy to adjust any other suggestions.

Akkadius commented:
Looks like this needs to be rebased with master

@hgtw any last things from your review?

Akkadius commented on common/eq_constants.h:
```diff
@@ -1131,4 +1131,12 @@ namespace LeadershipAbilitySlot {
+#define PARCEL_SEND_ITEMS    0
+#define PARCEL_SEND_MONEY    1
+#define PARCEL_MONEY_ITEM_ID 99990 // item id of money
+#define PARCEL_MAX_ITEMS     RuleI(Parcel, ParcelMaxItems)
```
`PARCEL_MAX_ITEMS` is compile time

Rule value will not change dynamically

neckkola replied:
Updated this define.

Akkadius commented on world/world_event_scheduler.cpp:
```diff
@@ -57,12 +59,53 @@ void WorldEventScheduler::Process(ZSList *zs_list)
 					zs_list->SendPacket(pack);
 					safe_delete(pack);
 				}
+				if (RuleB(Parcel, EnableParcelMerchants) && RuleB(Parcel, EnablePruning) && e.event_type == ServerEvents::EVENT_TYPE_PARCEL_PRUNE) {
+					LogScheduler("Parcel Prune Event Reached [{}]", e.event_data);
```
Should this be a configurable scheduler event type? It seems like this should just run in the world process loop as its own pruner processor

Akkadius replied:
See similar to player events truncation, this should live similarly. I don't think this belongs in the server event scheduler

neckkola replied:
I moved the prune routine to a world loop process from the scheduler process.

Akkadius commented on common/database/database_update_manifest.cpp:
```diff
@@ -5468,8 +5468,38 @@ DROP PRIMARY KEY,
+				ALTER TABLE `npc_types`
+				ADD COLUMN `is_parcel_merchant` TINYINT(1) UNSIGNED NOT NULL DEFAULT '0' AFTER `keeps_sold_items`;
+
+				UPDATE `npc_types` SET `is_parcel_merchant` = 1, `lastname` = 'Parcels and General Supplies'
+				WHERE id IN (202129, 3036, 394025, 75113, 49073, 41021, 40070, 106115, 55150, 9053, 382156, 1032,
```
Are we sure Parcel merchants aren't just a different class type?

We could automatically set the last name based on the class type and drop the new field

```cpp
	constexpr uint8 WizardGM                  = 31;
	constexpr uint8 MagicianGM                = 32;
	constexpr uint8 EnchanterGM               = 33;
	constexpr uint8 BeastlordGM               = 34;
	constexpr uint8 BerserkerGM               = 35;
	constexpr uint8 Banker                    = 40;
	constexpr uint8 Merchant                  = 41;
	constexpr uint8 DiscordMerchant           = 59;
	constexpr uint8 AdventureRecruiter        = 60;
	constexpr uint8 AdventureMerchant         = 61;
	constexpr uint8 LDoNTreasure              = 62;
	constexpr uint8 TributeMaster             = 63;
	constexpr uint8 GuildTributeMaster        = 64;
	constexpr uint8 GuildBanker               = 66;
	constexpr uint8 NorrathsKeepersMerchant   = 67;
	constexpr uint8 DarkReignMerchant         = 68;
	constexpr uint8 FellowshipMaster          = 69;
	constexpr uint8 AlternateCurrencyMerchant = 70;
	constexpr uint8 MercenaryLiaison          = 71;
```

neckkola commented:
Good afternoon.  I updated based on recent comments and rebased.  Still need to resolve the content aspects for merchant ids.  Let me know if I can assist with that in any way or if there are further comments.
